src/bin/pg_autoctl/fsm.c

```diff
@@ -287,6 +324,32 @@ KeeperFSMTransition KeeperFSM[] = {
 	{ PRIMARY_STATE, APPLY_SETTINGS_STATE, COMMENT_PRIMARY_TO_APPLY_SETTINGS, &fsm_apply_settings },
 	{ APPLY_SETTINGS_STATE, PRIMARY_STATE, COMMENT_APPLY_SETTINGS_TO_PRIMARY, NULL },
 
+	/*
+	 * In case of multiple standbys, failover begins with reporting current LSN
+	 */
+	{ SECONDARY_STATE, REPORT_LSN_STATE, COMMENT_SECONDARY_TO_REPORT_LSN, &fsm_report_lsn },
+	{ CATCHINGUP_STATE, REPORT_LSN_STATE, COMMENT_SECONDARY_TO_REPORT_LSN, &fsm_report_lsn },
+	{ REPORT_LSN_STATE, PREP_PROMOTION_STATE, COMMENT_REPORT_LSN_TO_PREP_PROMOTION, &fsm_prepare_standby_for_promotion },
+
+	{ REPORT_LSN_STATE, WAIT_FORWARD_STATE, COMMENT_REPORT_LSN_TO_WAIT_FORWARD, NULL },
```

Comment:
The reason why we have WAIT_CASCADE and WAIT_FORWARD is to orchestrate the addition of the failover candidate in the HBA of the most advanced standby, so that it may then proceed and connect to a new upstream server and fetch the missing bits from the WAL. We might be able to remove those two extra steps when we merge https://github.com/citusdata/pg_auto_failover/pull/311, but that work has not been done yet.

In the FSM structure a NULL pointer to a transition function means that we can transition and have nothing specific to do, as per the following code in `keeper_fsm_reach_assigned_state`:

``` c
			if (transition.transitionFunction)
			{
				ret = (*transition.transitionFunction)(keeper);

				log_debug("Transition function returned: %s",
						  ret ? "true" : "false");
			}
			else
			{
				ret = true;
				log_debug("No transition function, assigning new state");
			}
```

Reply:
We removes those two states now. I keep wondering about the “most advanced LSN standby node” choice, that we used to implement on the monitor and that is now done client-side with a SQL query to the monitor. It might be a good thing, and I'm going to add a restriction `where health = 1` to that query.